[Docs] Server Testing Guidelines (#3297)
ls1intum/Artemis: docs/dev/guidelines/server.rst

Some parts of these guidelines are adapted from https://medium.com/@madhupathy/ultimate-clean-code-guide-for-java-spring-based-applications-4d4c9095cc2a

22. General Testing Tips
========================

Write meaningful comments for your tests.
These comments should contain information about what is tested specifically.

.. code-block:: java

    /**
     * Tests that borrow() in Book successfully sets the available attribute to false
     */
    @Test
    public void testBorrowInBook() {
        // Test Code
    }

Use appropriate and descriptive names for test cases. This makes it easier for other developers to understand what you actually test without looking deeper into it.

> This is the same reason as to why you should not name your variables int a, double b, String c, and so on. For example, if you want to test the method borrow in the class Book, testBorrowInBook() would be an appropriate name for the test case.

Reviewer: is as to necessary here?

.. code-block:: java

    @Test
    public void testBorrowInBook() {
        // Test Code
    }

> Try to follow the best practices for java Testing:

Reviewer 2: ```suggestion
Try to follow the best practices for Java testing:
```

* Write small and specific tests by heavily using helper functions, parameterized tests, AssertJ’s powerful assertions, not overusing variables, asserting only what’s relevant and avoiding one test for all corner cases.
* Write self-contained tests by revealing all relevant parameters, insert data right in the test and prefer composition over inheritance.
* Write dumb tests by avoiding the reuse of production code and focusing on comparing output values with hard-coded values.
* KISS > DRY ("Keep it simple, Stupid!" and "Don't repeat yourself!")
* Test close to production by focusing on testing a complete vertical slide and avoiding in-memory databases.

> * JUnit5 and AssertJ are a very good choice.

Reviewer 2: Tbh I don't know if we need to mention those two here, since the project already requires tests to use AssertJ for example. I also don't know what is meant by "avoiding in-memory databases". Maybe we can drop that too since we have specific database utils?

Author: you are right I guess

For a more detail overview definitely check out: https://phauer.com/2019/modern-best-practices-testing-java/

Make use of JUnit 5 Features:
https://junit.org/junit5/docs/current/user-guide/#writing-tests
https://junit.org/junit5/docs/current/api/org.junit.jupiter.api/org/junit/jupiter/api/Assertions.html

Here you can find JUnit 5 best practices:
https://howtodoinjava.com/best-practices/unit-testing-best-practices-junit-reference-guide/

Also check out this page for spring related testing:
https://www.baeldung.com/spring-tests

If you want to write tests for Programming Exercises to test student's submission check out `this <https://confluence.ase.in.tum.de/display/ArTEMiS/Best+Practices+for+writing+Java+Programming+Exercise+Tests+in+Artemis>`__.

23. Avoid using @MockBean
=========================

Do not use the ``@SpyBean`` or ``@MockBean`` annotation unless absolutely necessary, or possibly in an abstract Superclass. If you want to see more in detail why, take a look `here <https://www.baeldung.com/spring-tests>`__.

> Basically every time ``@MockBean`` appears in a class, the ApplicationContext cache gets marked as dirty, hence the runner will clean the cache after the test-class is done and restarts the application context.

Reviewer 3: ```suggestion
Basically, every time ``@MockBean`` appears in a class, the ApplicationContext cache gets marked as dirty, hence the runner will clean the cache after the test-class is done and restarts the application context.
```

Author: looks better :+1:

> This leads to a large overhead, which makes the tests tend to take a lot more time.

Reviewer 3: ```suggestion
This leads to a large overhead, which tends to make the tests take a lot more time.
```

Here is an example how to replace a ``@SpyBean``. We wanted to test a catch case which is only executed if an IOException gets thrown. We did this by mocking the service method and making it throw an Exception.

.. code-block:: java

    public class TestExport extends AbstractSpringIntegrationBambooBitbucketJiraTest {
        @SpyBean
        private FileUploadSubmissionExportService fileUploadSubmissionExportService;

        @Test
        @WithMockUser(username = "instructor1", roles = "INSTRUCTOR")
        public void testExportAll_IOException() throws Exception {
            doThrow(IOException.class).when(fileUploadSubmissionExportService).export(any(), any());
            request.postWithResponseBodyFile("/api/file-upload-export/" + fileUploadExercise.getId()", HttpStatus.BAD_REQUEST);
        }
    }

> This, as mentioned above, we should really avoid.

Reviewer: As mentioned above, we should really avoid this

Author: I changed it :+1:

Instead we can use `Static Mocks <https://asolntsev.github.io/en/2020/07/11/mockito-static-methods/>`_. When we look deeper in the ``export()`` method we find that there is a call of ``File.newOutputStream(..)``.
Now, instead of mocking the whole Service we can just mock this static method, like this:

.. code-block:: java

    public class TestExport extends AbstractSpringIntegrationBambooBitbucketJiraTest {
        // No beans used anymore
        @Test
        @WithMockUser(username = "instructor1", roles = "INSTRUCTOR")
        public void testExportAll_IOException() throws Exception {
            MockedStatic<Files> mockedFiles = mockStatic(Files.class);
            mockedFiles.when(() -> Files.newOutputStream(any(), any())).thenThrow(IOException.class);
            request.postWithResponseBodyFile("/api/file-upload-export/" + fileUploadExercise.getId()", HttpStatus.BAD_REQUEST);

            mockedFiles.close();
        }
    }

> What you notice here is that we can avoid the use of a Bean and also test deeper. Instead of mocking the uppermost method we will only throw the exception at the place where it could actually happen. Very important to mention is that you need to close the mock at the end of the test again.

Reviewer: What you should notice, we can avoid ...

Some parts of these guidelines are adapted from https://medium.com/@madhupathy/ultimate-clean-code-guide-for-java-spring-based-applications-4d4c9095cc2a

> 22. General Testing Tips

Reviewer 2: Most of this paragraph reads like it is detailing testing tips for writing tests for exercises (i.e. which give students feedback). I'm not sure if the `dev` documentation is the right place for that.

.. code-block:: java

    /**
     * Tests that borrow() in Book successfully sets the available attribute to false
     * Problem Statement Task 2.1
     * Worth 1.5 Points (Weight: 1)
     */
    @Test
    public void testBorrowInBook() {
        // Test Code
    }

Better yet, for manual correction, make it the display name of the test. This allows the assessors to execute the tests in the IDE, have more meaningful names displayed and count points better.

.. code-block:: java

    @DisplayName("1.5 P | Books can be borrowed successfully")
    @Test
    public void testBorrowInBook() {
        // Test Code
    }

Use appropriate and descriptive names for test cases. After exams, test names will be used to create statistics. If the tests are called test1, test2, test3, and so on, no one can read those statistics.
This is the same reason as to why you should not name your variables int a, double b, String c, and so on. For example, if you want to test the method borrow in the class Book, testBorrowInBook() would be an appropriate name for the test case.

.. code-block:: java

    @Test
    public void testBorrowInBook() {
        // Test Code
    }

> se appropriate timeouts for test cases and take Bamboo into consideration. For regular test cases, using a ``@StrictTimeout(1)``  annotation is enough. This represents a strict timeout of one second.

Reviewer 4: ```suggestion
set appropriate timeouts for test cases and take Bamboo into consideration. For regular test cases, using a ``@StrictTimeout(1)``  annotation is enough. This represents a strict timeout of one second.
```

Reviewer 2: Probably upper-case, too?

> set appropriate timeouts for test cases and take Bamboo into consideration. For regular test cases, using a ``@StrictTimeout(1)``  annotation is enough. This represents a strict timeout of one second.

Reviewer 2: ```suggestion
Set appropriate timeouts for test cases and take Bamboo into consideration. For regular test cases, using a ``@StrictTimeout(1)`` annotation is enough. This represents a strict timeout of one second.
```

The value type of the strict timeout annotation defaults to seconds, if you need, for some reason, a shorter timeout, something like ``@StrictTimeout(value = 500, unit = TimeUnit.MILLISECONDS)`` works as well.

.. code-block:: java

    @Test
    @StrictTimeout(1)
    public void testBorrowInBook() {
        // Test Code
    }

> Avoiding assert statements and using conditional fail() calls instead hides confusing information from the students.

Reviewer 2: This sentence doesn't make sense to me. Maybe something like
```suggestion
Avoid assert statements and use conditional ``fail()`` calls instead to hide confusing information from the students.
```
?

This could be considered bad practice in regular conditions but helps to create fail messages that are less confusing, especially for beginners.
This also hides information about test implementation details.

.. code-block:: java

    @Test
    public void testBorrowInBook() {
        Object book = newInstance("Book", 0, "Some title");
        invokeMethod(book, "borrow");
        assertFalse((Boolean) invokeMethod(book, "isAvailable"), "A borrowed book must be unavailable!");
    }

If the student fails the test, Artemis will display something like org.opentest4j.AssertionFailedError:

> A borrowed book must be unavailable! ==> Expected <false> but was <true>. The part after '==>' should not be shown to the student as it contains implementation detail.

Reviewer 2: ```suggestion
A borrowed book must be unavailable! ==> Expected <false> but was <true>. The part after '==>' should not be shown to the student as it contains implementation details.
```

        invokeMethod(book, "borrow");
        if ((Boolean) invokeMethod(book, "isAvailable")) {
            fail("A borrowed book must be unavailable!");
        }
    }

This will lead to an error message like 'The class 'Book' was not found within the submission. Make sure to implement it properly.' Which is clear and tells the student exactly what is wrong with their code.

Some courses use long package identifiers like 'de.tum.in.ase.pse'. When instantiating objects with reflections,
the instantiation method usually takes the full canonical name of a class, which is 'de.tum.in.ase.pse.Book' for example.
To avoid writing out this full canonical name all the time, you can add a constant String attribute representing the base package name at the top of your test class

.. code-block:: java

    private static final String BASE_PACKAGE = "de.tum.in.ase.pse.";

    @Test
    public void testBorrowInBook() {
        Object book = newInstance(BASE_PACKAGE + "Book", 0, "Some title");
        // Test Code
    }

It is possible that students hardcode values to pass a certain set of tests, therefore you should check whether this is the case or not.

> This is especially important in an exam setting, so students don't get awarded points for a solution that does not fulfill the requirements described in the problem statement. (TODO: Code example)

Reviewer 2: This todo probably wants to be resolved.

        }
        String expected = "Dog:French";
        String actual = testTranslationServer.runService("Dog", "French");
        if(!expected.equals(actual)) fail("Descriptive fail message");
    }

This test correctly tests if the student added the case for the French language and called the appropriate method with the appropriate parameters. Because the translation method was overridden,
it doesn't matter whether the student has already completed the previous task or not.
(If you do that, you should have some way to deal with students that make the class or method final, either via problem statement or test, otherwise students get compilation errors in the test code)

Make use of JUnit 5 Features:
https://junit.org/junit5/docs/current/user-guide/#writing-tests
https://junit.org/junit5/docs/current/api/org.junit.jupiter.api/org/junit/jupiter/api/Assertions.html

General Testing best practices are found on https://phauer.com/2019/modern-best-practices-testing-java/

Also check out https://www.baeldung.com/spring-tests

And https://howtodoinjava.com/best-practices/unit-testing-best-practices-junit-reference-guide/

23. Avoid using @MockBean
=========================

> Do not use the ``@SpyBean`` or ``@MockBean`` annotation unless absolutely necessary, or possibly in an abstract Superclass. If you want to see more in detail why, take a look `here. <https://www.baeldung.com/spring-tests>`_

Reviewer 2: ```suggestion
Do not use the ``@SpyBean`` or ``@MockBean`` annotation unless absolutely necessary, or possibly in an abstract Superclass. If you want to see more in detail why, take a look `here <https://www.baeldung.com/spring-tests>`__.
```

> Basically every time @MockBean appears in a class, the ApplicationContext cache gets marked as dirty, hence the runner will clean the cache after the test-class is done and restarts the application context.

Reviewer 5: You could unify the use of ``@MockBean`` and @Mockbean

Author: ```suggestion
Basically every time ``@MockBean`` appears in a class, the ApplicationContext cache gets marked as dirty, hence the runner will clean the cache after the test-class is done and restarts the application context.
```
like this you mean?

This leads to a large overhead, which makes the tests tend to take a lot more time.

Here is an example how to replace a ``@SpyBean``. We wanted to test a catch case which is only executed if an IOException gets thrown. We did this by mocking the service method and making it throw an Exception.

.. code-block:: java

    public class TestExport extends AbstractSpringIntegrationBambooBitbucketJiraTest {
        @SpyBean
        private FileUploadSubmissionExportService fileUploadSubmissionExportService;

        @Test
        @WithMockUser(username = "instructor1", roles = "INSTRUCTOR")
        public void testExportAll_IOException() throws Exception {
            doThrow(IOException.class).when(fileUploadSubmissionExportService).export(any(), any());
            request.postWithResponseBodyFile("/api/file-upload-export/" + fileUploadExercise.getId()", HttpStatus.BAD_REQUEST);
        }
    }

> This, as mentioned above, we should really avoid this.

Reviewer 5: 2x this is not intended, is it?

Author: Nope that's not intended
```suggestion
This, as mentioned above, we should really avoid.
```

> Instead we can use Static Mocks. When we look deeper in the ``export()`` method we find that there is a call of ``File.newOutputStream(..)``.

Reviewer 5: Maybe a link to more information about static mocks would be helpful?

Author: ```suggestion
Instead we can use `Static Mocks <https://asolntsev.github.io/en/2020/07/11/mockito-static-methods/>`_. When we look deeper in the ``export()`` method we find that there is a call of ``File.newOutputStream(..)``.
```

Now, instead of mocking the whole Service we can just mock this static method, like this:

.. code-block:: java

    public class TestExport extends AbstractSpringIntegrationBambooBitbucketJiraTest {
        // No beans used anymore
        @Test
        @WithMockUser(username = "instructor1", roles = "INSTRUCTOR")
        public void testExportAll_IOException() throws Exception {
            MockedStatic<Files> mockedFiles = mockStatic(Files.class);
            mockedFiles.when(() -> Files.newOutputStream(any(), any())).thenThrow(IOException.class);
            request.postWithResponseBodyFile("/api/file-upload-export/" + fileUploadExercise.getId()", HttpStatus.BAD_REQUEST);

            mockedFiles.close();
        }
    }

> What you notice here is that we can avoid the use of a Bean and also test deeper. Important to mention is that you need to close the mock at the end of the test again.

Reviewer 5: Maybe a sentence on "why this is deeper" (i.e. what is now covered which was not before with a Bean?) would be good.

Author: ```suggestion
What you notice here is that we can avoid the use of a Bean and also test deeper. Instead of mocking the uppermost method we will only throw the exception at the place where it could actually happen. Very important to mention is that you need to close the mock at the end of the test again.
```

> For a real examples where a SpyBean was replace with a static mock look `here. <https://github.com/ls1intum/Artemis/commit/4843137aa01cfdf27ea019400c48df00df36ed45>`_

Reviewer 5: You could unify to ``@SpyBean` and capitalize Static Mock as above.

Reviewer 5: This is not really different to the in-doc example, is it necessary? If you want to keep it, I would add the file name (an even lines?) to search for the replacement example. So readers do not have to browse through all the changed files. I think you cannot link the exact lines unfortunately.

Reviewer 5: typo: replace_d_

Author: ```suggestion
For a real example where a SpyBean was replaced with a static mock look at the SubmissionExportIntegrationTest.java in `here. <https://github.com/ls1intum/Artemis/commit/4843137aa01cfdf27ea019400c48df00df36ed45>`_
```

Author: good catch :+1:

> For a real example where a SpyBean was replaced with a static mock look at the SubmissionExportIntegrationTest.java in `here. <https://github.com/ls1intum/Artemis/commit/4843137aa01cfdf27ea019400c48df00df36ed45>`_

Reviewer 2: ```suggestion
For a real example where a SpyBean was replaced with a static mock look at the SubmissionExportIntegrationTest.java in `here <https://github.com/ls1intum/Artemis/commit/4843137aa01cfdf27ea019400c48df00df36ed45>`__.
```

This is the same reason why you should not name your variables int a, double b, String c, and so on. For example, if you want to test the method borrow in the class Book, testBorrowInBook() would be an appropriate name for the test case.

.. code-block:: java

    @Test
    public void testBorrowInBook() {
        // Test Code
    }

Try to follow the best practices for Java testing:

* Write small and specific tests by heavily using helper functions, parameterized tests, AssertJ’s powerful assertions, not overusing variables, asserting only what’s relevant and avoiding one test for all corner cases.
* Write self-contained tests by revealing all relevant parameters, insert data right in the test and prefer composition over inheritance.
* Write dumb tests by avoiding the reuse of production code and focusing on comparing output values with hard-coded values.
* KISS > DRY ("Keep it simple, Stupid!" and "Don't repeat yourself!")
* Invest in a testable implementation by avoiding static access, using constructor injection, using Clocks and separating business logic from asynchronous execution.

> For a more detail overview definitely check out: https://phauer.com/2019/modern-best-practices-testing-java/

Reviewer 6: ```suggestion
For a more detailed overview definitely check out: https://phauer.com/2019/modern-best-practices-testing-java/
```

Make use of JUnit 5 Features:
https://junit.org/junit5/docs/current/user-guide/#writing-tests
https://junit.org/junit5/docs/current/api/org.junit.jupiter.api/org/junit/jupiter/api/Assertions.html

Here you can find JUnit 5 best practices:
https://howtodoinjava.com/best-practices/unit-testing-best-practices-junit-reference-guide/

Also check out this page for spring related testing:
https://www.baeldung.com/spring-tests

If you want to write tests for Programming Exercises to test student's submissions check out `this <https://confluence.ase.in.tum.de/display/ArTEMiS/Best+Practices+for+writing+Java+Programming+Exercise+Tests+in+Artemis>`__.

23. Avoid using @MockBean
=========================

> Do not use the ``@SpyBean`` or ``@MockBean`` annotation unless absolutely necessary, or possibly in an abstract Superclass. If you want to see more in detail why, take a look `here <https://www.baeldung.com/spring-tests>`__.

Reviewer 3: ```suggestion
Do not use the ``@SpyBean`` or ``@MockBean`` annotation unless absolutely necessary, or possibly in an abstract Superclass. If you want to see why in more detail, take a look `here <https://www.baeldung.com/spring-tests>`__.
```

Basically, every time ``@MockBean`` appears in a class, the ApplicationContext cache gets marked as dirty, hence the runner will clean the cache after the test-class is done and restarts the application context.
This leads to a large overhead, which tends to make the tests take a lot more time.

> Here is an example how to replace a ``@SpyBean``. We wanted to test a catch case which is only executed if an IOException gets thrown. We did this by mocking the service method and making it throw an Exception.

Reviewer 6: ```suggestion
Here is an example how to replace a ``@SpyBean``. We wanted to test a catch case which is only executed if an IOException is thrown. We did this by mocking the service method and making it throw an Exception.
```

.. code-block:: java

    public class TestExport extends AbstractSpringIntegrationBambooBitbucketJiraTest {
        @SpyBean
        private FileUploadSubmissionExportService fileUploadSubmissionExportService;

        @Test
        @WithMockUser(username = "instructor1", roles = "INSTRUCTOR")
        public void testExportAll_IOException() throws Exception {
            doThrow(IOException.class).when(fileUploadSubmissionExportService).export(any(), any());
            request.postWithResponseBodyFile("/api/file-upload-export/" + fileUploadExercise.getId()", HttpStatus.BAD_REQUEST);
        }
    }

> As mentioned above, we should really avoid this

Reviewer 3: ```suggestion
As mentioned above, we should really avoid this.
```

Instead we can use `Static Mocks <https://asolntsev.github.io/en/2020/07/11/mockito-static-methods/>`_. When we look deeper in the ``export()`` method we find that there is a call of ``File.newOutputStream(..)``.

> Now, instead of mocking the whole Service we can just mock this static method, like this:

Reviewer 6: ```suggestion
Now, instead of mocking the whole Service, we can just mock this static method, like this:
```

.. code-block:: java

    public class TestExport extends AbstractSpringIntegrationBambooBitbucketJiraTest {
        // No beans used anymore
        @Test
        @WithMockUser(username = "instructor1", roles = "INSTRUCTOR")
        public void testExportAll_IOException() throws Exception {
            MockedStatic<Files> mockedFiles = mockStatic(Files.class);
            mockedFiles.when(() -> Files.newOutputStream(any(), any())).thenThrow(IOException.class);
            request.postWithResponseBodyFile("/api/file-upload-export/" + fileUploadExercise.getId()", HttpStatus.BAD_REQUEST);

            mockedFiles.close();
        }
    }

> You should notice here that we can avoid the use of a Bean and also test deeper. Instead of mocking the uppermost method we will only throw the exception at the place where it could actually happen. Very important to mention is that you need to close the mock at the end of the test again.

Reviewer 3: ```suggestion
You should notice here that we can avoid the use of a Bean and also test deeper. Instead of mocking the uppermost method we only throw the exception at the place where it could actually happen. Very important to mention is that you need to close the mock at the end of the test again.
```

Reviewer 6: The `will` doesn't feel wrong here to me, but it's a matter of taste I suppose :P
